org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java

```diff
@@ -925,25 +933,79 @@ public static void applyWorkspaceEdit(WorkspaceEdit wsEdit, String label) {
 		String name = label == null ? DEFAULT_LABEL : label;
 		Map<URI, Range> changedURIs = new HashMap<>();
 		CompositeChange change = toCompositeChange(wsEdit, name, changedURIs);
-		final var changeOperation = new PerformChangeOperation(change);
-		changeOperation.setUndoManager(RefactoringCore.getUndoManager(), name);
-		try {
-			ResourcesPlugin.getWorkspace().run(changeOperation, new NullProgressMonitor());
-
-			// Open the resource in editor if there is the only one URI
-			if (changedURIs.size() == 1) {
-				changedURIs.keySet().stream().findFirst().ifPresent(uri -> {
-					// Select the only start position of the range or the document start
-					Range range = changedURIs.get(uri);
-					Position start = range.getStart() != null ? range.getStart() : new Position(0, 0);
-					if (Display.getCurrent() != null) {
-						open(uri.toString(), new Range(start, start));
-					} else {
-						UI.getDisplay().asyncExec(() -> open(uri.toString(), new Range(start, start)));
-					}
-				});
+
+		if (wsEdit.getChangeAnnotations() != null && wsEdit.getChangeAnnotations().values().stream().anyMatch(ca -> ca.getNeedsConfirmation())) {
+			Refactoring refactoring = new Refactoring() {
```

Comment:
Could we move lines 938-973 into `runRefactorWizardOperation` passing the `change` to that method instead of the `wizard` as now?

Reply:
Yes, definitely!